ipyparallel/cluster/shellcmd.py

```diff
@@ -0,0 +1,860 @@
+        )  # replace possible ${userdir} placeholder
+
+    def _output_msg(self, level, msg):
+        with open(self.filename, "a") as f:
+            f.write(f"{level:8} {datetime.now()} {msg}\n")
+
+    def info(self, msg):
+        self._output_msg("info", msg)
+
+    def warning(self, msg):
+        self._output_msg("warning", msg)
+
+    def error(self, msg):
+        self._output_msg("error", msg)
+
+    def debug(self, msg):
+        self._output_msg("debug", msg)
+
+
+class ShellCommandReceive:
+    """
+    Helper class for receiving and performing shell commands in a platform independent form
+
+    All supported shell commands have a cmd_ prefix. When adding new functions make sure that the is an
+    equivalent in the ShellCommandSend class. When a command failed a non zero exit code will be returned.
+    Hence, the ShellCommandSend class always uses subprocess.check_output for assessing if the command was
+    successfull. Some command require information to be returned (cmd_exists, cmd_start, cmd_running) which
+    is written to stdout in the following form: __<key>=<value>__
+    """
+
+    def _log(self, msg):
+        if not self.log:
+            return
+        self.log.info(f"[id={self.ranid}] {msg}")
+
+    def __init__(self, debugging=False, use_break_way=True, log=None):
+        self.debugging = debugging
+        self.log = None
+        self.platform = Platform.get()
+        self.use_break_way = use_break_way
+        if log:
+            if isinstance(log, str):
+                self.log = SimpleLog(log)
+            else:
+                self.log = log
+        elif "SHELLCMD_LOG" in os.environ:
+            self.log = SimpleLog(os.environ["SHELLCMD_LOG"])
+
+        self.ranid = None
+        if self.log:
+            self.ranid = randint(0, 999)
+            self._log("ShellCommandReceive instance created")
+
+    def __del__(self):
+        if self.log:
+            self._log("ShellCommandReceive instance deleted")
+
+    def _linux_quote(self, p):
+        if "'" in p:
+            return '"' + p + '"'
+        else:
+            return p
+
+    def cmd_start(self, start_cmd, env=None, output_file=None):
+        if env:
+            # add provided entries to environment
+            if isinstance(env, str):
+                if env[0] == '"' and env[-1] == '"':
+                    env = env.strip('"')  # occurs under windows cmd
+                if "\\'" in env:
+                    env = env.replace("\\'", "'")  # replace quoted
+                env_dict = eval(env)
+                assert isinstance(env_dict, dict) is True
+                env = env_dict
+
+            assert isinstance(env, dict) is True  # make sure that env is a dictionary
+
+            # update environment
+            for key, value in env.items():
+                if value is not None and value != '':
+                    if isinstance(value, dict):
+                        os.environ[key] = json.dumps(value)
```

Comment:
On the recieving side, we should assume/require that the env dict is a valid enviornment dict. There should be no dictionaries or other non-string types on the right hand side.

Reply:
This was a sort of convenience functionality so that you can directly pass a dict of dict as env (as conceptionally used for IPP_CONNECTION_INFO). But I think you are right, it’s maybe more confusing than helping. I have removed the automatic conversion and also added a [check](https://github.com/ottointhesky/ipyparallel/blob/0361b687bc7fdd8f9259e9478abf424dab894aa2/ipyparallel/cluster/shellcmd.py#L467C1-L479C1) to make sure that the env dict only contains strings (values and keys).

This also effected the code in [test_shellcmd.py](https://github.com/ottointhesky/ipyparallel/blob/0361b687bc7fdd8f9259e9478abf424dab894aa2/ipyparallel/tests/test_shellcmd.py#L234C1-L257C6)